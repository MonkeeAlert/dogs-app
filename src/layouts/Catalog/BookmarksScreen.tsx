import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  NativeSyntheticEvent,
  NativeScrollEvent,
} from 'react-native';
import React, {useEffect, useState} from 'react';
import {useSelector} from 'react-redux';
import {getDogsCatalog} from '../../redux/rootSelector';
import {colors, text} from '../../utils/constants';
import CustomStatusBar from '../../components/CustomStatusBar';
import {parseImage} from '../../utils/functions';
import DogImageListItem from '../../components/lists/DogImageListItem';
import EmptyList from '../../components/lists/EmptyList';
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedStyle,
  useSharedValue,
} from 'react-native-reanimated';
import {Icon} from 'react-native-elements/dist/icons/Icon';
import {MainStyles} from '../../assets/styles/MainStyles';

const renderItem = (uri: string, idx: number) => {
  return <DogImageListItem uri={uri} idx={idx} />;
};

const BookmarksScreen = () => {
  const {bookmarks} = useSelector(getDogsCatalog);
  const [data, setData] = useState<string[]>([]);

  const scrollY = useSharedValue(0);

  const counterStyle = useAnimatedStyle(() => ({
    opacity: interpolate(scrollY.value, [0, 49], [0, 1], Extrapolate.CLAMP),
    transform: [
      {
        translateY: interpolate(
          scrollY.value,
          [0, 49],
          [10, 0],
          Extrapolate.CLAMP,
        ),
      },
    ],
  }));

  const handleScroll = (e: NativeSyntheticEvent<NativeScrollEvent>) => {
    scrollY.value = e.nativeEvent.contentOffset.y;
  };

  // scroll handler for FlatList
  useEffect(() => {
    if (bookmarks.length > 4 && bookmarks.length < 7) {
      setData([...bookmarks, '', '']);
    } else {
      setData(bookmarks);
    }
  }, [bookmarks]);

  return (
    <View style={styles.container}>
      <CustomStatusBar
        backgroundColor={colors.turquoise}
        barStyle={'dark-content'}
      />
      <View style={styles.header}>
        <Text style={styles.headerText}>Favourites:</Text>

        <View style={styles.counterContainer}>
          <Animated.View style={[styles.counter, counterStyle]}>
            <Text style={[styles.headerText, {fontSize: text.s}]}>
              {bookmarks.length}
            </Text>
            <Icon
              name={'bookmarks'}
              type={'ionicon'}
              size={text.m}
              color={colors.white}
            />
          </Animated.View>
        </View>
      </View>

      <Animated.FlatList
        testID={'BookmarksScreen_List'}
        data={data}
        numColumns={2}
        showsVerticalScrollIndicator={false}
        keyExtractor={parseImage}
        contentContainerStyle={[
          styles.list,
          data.length === 0
            ? styles.emptyList
            : data.length <= 6
            ? {height: '100%'}
            : {},
        ]}
        bounces={false}
        renderItem={({item, index}) => renderItem(item, index)}
        onScroll={handleScroll}
        scrollEventThrottle={16}
        ListEmptyComponent={EmptyList}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: '100%',
    backgroundColor: colors.turquoise,
    position: 'relative',
  },
  counterContainer: {
    width: 49,
    overflow: 'hidden',
  },
  counter: {
    ...MainStyles.rowFull,
  },
  header: {
    ...MainStyles.rowFull,
    alignItems: 'flex-end',
    marginVertical: 7,
    marginHorizontal: 14,
  },
  headerText: {
    color: colors.white,
    fontSize: text.m,
    fontWeight: '900',
  },
  list: {
    paddingHorizontal: 7,
    paddingTop: 7,
    paddingBottom: 49,
    borderTopLeftRadius: 21,
    borderTopRightRadius: 21,
    backgroundColor: colors.white,
  },
  emptyList: {
    flex: 1,
    paddingBottom: 0,
    paddingTop: Dimensions.get('screen').height / 4,
  },
  indicator: {
    paddingTop: 49,
    marginBottom: 7,
  },
});

export default BookmarksScreen;
