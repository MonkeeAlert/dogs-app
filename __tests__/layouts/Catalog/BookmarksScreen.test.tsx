import {act, fireEvent, render} from '@testing-library/react-native';
import {shallow} from 'enzyme';
import React from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {create} from 'react-test-renderer';
import {mockedDispatch} from '../../../jest.setup';
import BookmarksScreen from '../../../src/layouts/Catalog/BookmarksScreen';

const LIST = [
  'https://images.dog.ceo/breeds/labrador/n02099712_5648.jpg',
  'https://images.dog.ceo/breeds/retriever-golden/n02099601_8005.jpg',
  'https://images.dog.ceo/breeds/pyrenees/n02111500_7483.jpg',
];

describe('BookmarksScreen', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(mockedDispatch);
  });

  it('matches snapshot', () => {
    useSelector.mockReturnValue({
      bookmarks: LIST,
    });

    act(() => {
      const tree = create(<BookmarksScreen />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('renders with list of 5 items', () => {
    useSelector.mockReturnValue({
      bookmarks: [
        ...LIST,
        'https://images.dog.ceo/breeds/labrador/n02090712_5648.jpg',
        'https://images.dog.ceo/breeds/labrador/n82099712_5648.jpg',
      ],
    });

    const {getAllByTestId} = render(<BookmarksScreen />);
    const items = getAllByTestId('DogImageListItem_View');

    expect(items.length).toBe(5);
  });

  it('handles onScroll', () => {
    useSelector.mockReturnValue({
      bookmarks: [
        ...LIST,
        'https://images.dog.ceo/breeds/labrador/n02090712_5648.jpg',
        'https://images.dog.ceo/breeds/labrador/n82099712_5648.jpg',
      ],
    });

    const wrapper = shallow(<BookmarksScreen />);
    const list = wrapper.find({testID: 'BookmarksScreen_List'});

    list.simulate('scroll', {nativeEvent: {contentOffset: {y: 1}}});
  });
});
