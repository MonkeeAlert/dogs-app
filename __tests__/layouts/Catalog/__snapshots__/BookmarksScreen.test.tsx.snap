// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarksScreen matches snapshot 1`] = `null`;
